clients/da-vinci-client/src/main/java/com/linkedin/davinci/kafka/consumer/ActiveActiveStoreIngestionTask.java

```diff
@@ -825,6 +825,97 @@ protected void produceToLocalKafka(
     }
   }
 
+  @Override
+  protected Map<String, Long> calculateLeaderUpstreamOffsetWithTopicSwitch(
+      PartitionConsumptionState partitionConsumptionState,
+      TopicSwitch topicSwitch,
+      PubSubTopic newSourceTopic,
+      List<CharSequence> unreachableBrokerList) {
+    Map<String, Long> upstreamOffsetsByKafkaURLs = new HashMap<>(topicSwitch.sourceKafkaServers.size());
+    final PubSubTopicPartition sourceTopicPartition = partitionConsumptionState.getSourceTopicPartition(newSourceTopic);
+    long rewindStartTimestamp;
+    // calculate the rewind start time here if controller asked to do so by using this sentinel value.
+    if (topicSwitch.rewindStartTimestamp == REWIND_TIME_DECIDED_BY_SERVER) {
+      rewindStartTimestamp = calculateRewindStartTime(partitionConsumptionState);
+      LOGGER.info(
+          "{} leader calculated rewindStartTimestamp {} for {}",
+          ingestionTaskName,
+          rewindStartTimestamp,
+          sourceTopicPartition);
+    } else {
+      rewindStartTimestamp = topicSwitch.rewindStartTimestamp;
+    }
+
+    topicSwitch.sourceKafkaServers.forEach(sourceKafkaURL -> {
+      Long upstreamStartOffset =
+          partitionConsumptionState.getLatestProcessedUpstreamRTOffsetWithNoDefault(sourceKafkaURL.toString());
+      if (upstreamStartOffset == null || upstreamStartOffset < 0) {
+        if (rewindStartTimestamp > 0) {
+          PubSubTopicPartition newSourceTopicPartition =
+              new PubSubTopicPartitionImpl(newSourceTopic, sourceTopicPartition.getPartitionNumber());
+          try {
+            upstreamStartOffset =
+                getTopicPartitionOffsetByKafkaURL(sourceKafkaURL, newSourceTopicPartition, rewindStartTimestamp);
+            LOGGER.info(
+                "Get upstreamStartOffset: {} for source URL: {}, topic: {}, rewind timestamp: {}",
+                upstreamStartOffset,
+                sourceKafkaURL,
+                newSourceTopicPartition,
+                rewindStartTimestamp);
+          } catch (Exception e) {
+            /**
+             * This is actually tricky. Potentially we could return a -1 value here, but this has the gotcha that if we
+             * have a non-symmetrical failure (like, lor1 can't talk to the ltx1 broker) this will result in a remote
```

Comment:
Let's replace these names with `region1`, `region2` as it doesn't make sense outside of LI...